src/nnet3/nnet-computation.cc

```diff
@@ -413,6 +608,210 @@ void NnetComputation::Print(std::ostream &os, const Nnet &nnet) const {
+    }
+  }
+
+  size_t num_input_output_info;
+  ExpectToken(is, binary, "<NumInputOutputInfo>");
+  ReadBasicType(is, binary, &num_input_output_info);
+  KALDI_ASSERT(num_input_output_info >= 0);
+  input_output_info.clear();
+  ExpectToken(is, binary, "<InputOutputInfo>");
+  for (size_t c = 0; c < num_input_output_info; c++) {
+    int32 key;
+    std::pair<int32, int32> val;
+    ReadBasicType(is, binary, &key);
+    ReadBasicType(is, binary, &(val.first));
+    ReadBasicType(is, binary, &(val.second));
+    input_output_info.insert(std::make_pair<int32, std::pair<int32, int32> >(key, val));
+  }
+
+  size_t num_commands;
+  ExpectToken(is, binary, "<NumCommands>");
+  ReadBasicType(is, binary, &num_commands);
+  KALDI_ASSERT(num_commands >= 0);
+  commands.resize(num_commands);
+  ExpectToken(is, binary, "<Commands>");
+  for (size_t c = 0; c < num_commands; c++) {
+    commands[c].Read(is, binary);
+  }
+
+  ExpectToken(is, binary, "<NeedModelDerivative>");
+  ReadBasicType(is, binary, &need_model_derivative);
+
+  ComputeCudaIndexes();
+}
+
+void NnetComputation::Write(std::ostream &os, bool binary) const {
+  WriteToken(os, binary, "<NumMatrices>");
+  WriteBasicType(os, binary, matrices.size());
+  WriteToken(os, binary, "<Matrices>");
+  for (size_t c = 0; c < matrices.size(); c++) {
+    matrices[c].Write(os, binary);
+  }
+
+  WriteToken(os, binary, "<NumMatrixDebugInfo>");
+  WriteBasicType(os, binary, matrix_debug_info.size());
+  WriteToken(os, binary, "<MatrixDebugInfo>");
+  for (size_t c = 0; c < matrix_debug_info.size(); c++) {
+    matrix_debug_info[c].Write(os, binary);
+  }
+
+  WriteToken(os, binary, "<NumSubMatrices>");
+  WriteBasicType(os, binary, submatrices.size());
+  WriteToken(os, binary, "<SubMatrices>");
+  for (size_t c = 0; c < submatrices.size(); c++) {
+    submatrices[c].Write(os, binary);
+  }
+
+//  WriteToken(os, binary, "<NumComponentPrecomputedIndexes>");
+//  WriteBasicType(os, binary, component_precomputed_indexes.size());
+//  WriteToken(os, binary, "<ComponentPrecomputedIndexes>");
+//  for (int32 c = 0; c < component_precomputed_indexes.size(); c++) {
+//    if (component_precomputed_indexes[c] != NULL) 
+//      component_precomputed_indexes[c]->Write(os, binary);
+//  }
+
+  WriteToken(os, binary, "<NumIndexes>");
+  WriteBasicType(os, binary, indexes.size());
+  WriteToken(os, binary, "<Indexes>");
+  for (size_t c = 0; c < indexes.size(); c++) {
+    WriteBasicType(os, binary, indexes[c].size());
+    for (size_t i = 0; i < indexes[c].size(); i++)
+      WriteBasicType(os, binary, indexes[c][i]);
+  }
+
+  WriteToken(os, binary, "<NumIndexesMulti>");
+  WriteBasicType(os, binary, indexes_multi.size());
+  WriteToken(os, binary, "<IndexesMulti>");
+  for (size_t c = 0; c < indexes_multi.size(); c++) {
+    WriteBasicType(os, binary, indexes_multi[c].size());
+    for (size_t i = 0; i < indexes_multi[c].size(); i++) {
+      WriteBasicType(os, binary, indexes_multi[c][i].first);
```

Comment:
You could add a function WriteIntegerPairVector modeled after WriteIntegerVector (assume the types are the same).